server/aesdsocket.c

```diff
@@ -0,0 +1,367 @@
+
+    // add the signals to the set
+    sigaddset(&set,SIGINT);      
+    sigaddset(&set,SIGTERM);     
+
+    
+    // lisiten for connections
+    check = listen(sock_t, BACKLOG);
+
+    if (check == -1){
+
+        perror("\nERROR listen():");
+        close_graceful();
+        exit(-1);
+
+    }
+
+    // allocate memory for buffers
+    read_buf = (char*)malloc(sizeof(char)*BUFSIZE);
+    write_buf = (char*)malloc(sizeof(char)*BUFSIZE);
+    
+    // if correct parameter is passed to code start daemon
+    if (argc == 2){
+
+        if (!strcmp("-d",argv[1])){
+
+        check = fork();
+        if (check == -1){
+
+            perror("\nERROR fork():");
+            close_graceful();
+            exit(-1);
+        }
+
+        if (check != 0){
+            exit(0);
+        }
+
+        setsid();
+
+        chdir("/");
+
+        open("/dev/null", O_RDWR);
+		dup(0);
+		dup(0);
+
+    }
+
+    }
+        
+
+
+    while(1) {
+
+
+    len = sizeof(con_addr);
+
+   // accept the connection
+    newsock_t = accept(sock_t,(struct sockaddr *)&con_addr,&len);
+
+    if (newsock_t == -1 ){
+        perror("\nERROR accept():");
+        //continue;
+        }
+
+
+    char *ip_string = inet_ntoa(con_addr.sin_addr);
+    //printf("\nAccepted connection from %s",ip_string);
+    syslog(LOG_DEBUG,"\nAccepted connection from %s",ip_string);
+    
+
+    char *ch;
+    ssize_t wbytes;
+
+    buff_pos=0;
+
+    // Read till packet is complete
+    while((num_bytes = recv(newsock_t,read_buf+buff_pos, BUFSIZE, 0))>0){
+
+    if(num_bytes == -1){
+        perror("ERROR recv():");
+        close_graceful();
+        exit(-1);    
+
+    }
+
+    buff_pos += num_bytes;
+    
+    // dynamicall increase buffer size for incoming packets
+    if (buff_pos >= currbuf_size){
+
+    currbuf_size+=BUFSIZE;
+
+    temp_buf = realloc(read_buf,sizeof(char)*currbuf_size);
+
+    if(temp_buf == NULL){
+        syslog(LOG_ERR,"\nErrror in realloc");
+        close_graceful();
+        exit(-1);
+    }
+
+    else read_buf = temp_buf;
+    
+    }
+    
+    // Search for NULL character
+    ch = strchr(read_buf,'\n');
+
+    if(ch != NULL) break;
+    
+    
+    }
+
+
+    // Block signals to avoid partial write
+    if (sigprocmask(SIG_BLOCK,&set,NULL) == -1){
+        perror("\nERROR sigprocmask():");
+        close_graceful();
+        exit(-1);
+    }
+
+    // Open file for writing
+    fd = open("/var/tmp/aesdsocketdata",O_RDWR|O_CREAT|O_APPEND,S_IRWXU);
+    if(fd<0){
+        perror("\nERROR open():");
+        close_graceful();
+        exit(-1);
+    }
+
+    // Write to file
+    wbytes = write(fd,read_buf,buff_pos);
+    if (wbytes == -1){
+        perror("\nERROR write():");
+        close_graceful();
+        exit(-1);
+    }
+    
+    // Extract file size
+    status = stat("/var/tmp/aesdsocketdata",&stat_data);
+    int si;
+    if (status == 0){
+    si=stat_data.st_size;
+
+    }
+
+    out_buf=realloc(write_buf,sizeof(char)*si);
```

Comment:
@shni9045 what assumption does this code make about the size of the RAM heap relative to the total size of all packets sent?